kube-runtime/src/scheduler.rs

```diff
@@ -138,29 +174,147 @@ where
     }
 }
 
+impl<T, R> Scheduler<T, R>
+where
+    T: Eq + Hash + Clone,
+    R: Stream<Item = ScheduleRequest<T>>,
+{
+    /// A filtered view of the [`Scheduler`], which will keep items "pending" if
+    /// `can_take_message` returns `false`, allowing them to be handled as soon as
+    /// they are ready.
+    ///
+    /// The returned [`HoldUnless`] is designed to be short-lived: it has no allocations, and
+    /// no messages will be lost, even if it is reconstructed on each call to [`poll_next`].
+    /// In fact, this is often desirable, to avoid long-lived borrows in `can_take_message`'s closure.
+    ///
+    /// NOTE: `can_take_message` should be considered fairly performance-sensitive, since
+    /// it will generally be executed for each pending message, for each [`poll_next`].
+    pub fn hold_unless<'a, C: Fn(&T) -> bool>(
```

Comment:
Knit: The `'a` lifetime could be elided by `rustc`, no need to be explicit, but doesn't hurt either :wink: .

Reply:
Yeah.. I guess it started getting into the complexity territory where I intuitively start writing them out explicitly.

Reply:
48da82f